assimilation_code/modules/utilities/utilities_mod.f90

```diff
@@ -492,6 +493,101 @@ subroutine check_namelist_read(iunit, iostat_in, nml_name)
 
 end subroutine check_namelist_read
 
+
+!-----------------------------------------------------------------------
+!> convert integers or strings describing options into integer. 
+!> 
+!> Some namelist items have historically been integers to describe inflation
+!> algorithms or ... which is a terrible thing to do to the users.
+!> This routine allows for backward compatibility and still provides
+!> the option to specify   inf_flavor = 'VARYING_SS_INFLATION' for example.
+
+function flex_parser(input,integer_options,string_options,context)
+
+character(len=*), intent(in) :: input                 ! value from namelist, eg
+integer,          intent(in) :: integer_options(:)    ! possible integer values
+character(len=*), intent(in) :: string_options(:)     ! matching string values
+character(len=*), intent(in), optional :: context 
+integer                      :: flex_parser
+
+character(len=len_trim(input)) :: uppercase
+character(len=len(string_options)) :: possibility
+integer :: ios, iopt, candidate
+
+if ( .not. module_initialized ) call initialize_utilities
+
+flex_parser = MISSING_I
+
+! Try to read the input as an ASCII-coded integer ( i.e. '3') 
+
+read(input,*,iostat=ios) candidate
+if (ios == 0) then
+
+   flex_parser = candidate
+
+   ! Check for valid value
+   if (any(integer_options == flex_parser)) then
+      return
+   else
+      call error_report(context)
+   endif
+endif
+
+! numeric conversion not possible, try to interpret as a string
+
+if (size(integer_options) /= size(string_options)) then
+   write(msgstring2,*)'size(integer_options) =',size(integer_options), &
+                  ' /= size( string_options) =',size( string_options)
+   call error_handler(E_ERR,'flex_parser',msgstring2,source,text2=context)
+endif
+
+uppercase = trim(input)
+call to_upper(uppercase)
+
+LOOP : do iopt = 1,size(integer_options)
+
+   possibility = string_options(iopt)
+   call to_upper(possibility)
+
+   if (uppercase == possibility) then
+      flex_parser = integer_options(iopt)
+      exit LOOP
+   endif
+
+enddo LOOP
+
+if (flex_parser == MISSING_I) call error_report(context)
+
+contains
```

Comment:
i like this code, but i'm not a fan of 'contains' in subroutines.  it confuses debuggers and profilers, and it prevents an embedded routine from being used anywhere else.  i realize this routine is specific to the flex_parser() code, but still i would move it to be a fully fledged subroutine but i wouldn't put it in the public list.  i might also rename it flex_parser_error() or something to tie it more closely to the parent routine.